dev-amdgpu,mem-ruby: Add support to checkpoint and restore between kernels in GPUFS

v-ramadas opened the pull request:
Earlier, GPU checkpointing was working only if a checkpoint was created before the first kernel execution. This pull request adds support to checkpoint in-between any two kernel calls. It does so by doing the following.

- Adds flush support in the GPU_VIPER protocol
- Adds flush support in the GPUCoalescer
- Updates cache recorder to use the GPUCoalescer during simulation cooldown and cache warmup times.

v-ramadas committed:
dev-amdgpu: Add aql, hsa queue information to checkpoint-restore

GPUFS uses aql information from PM4 queues to initialize doorbells. This
commit adds aql information to the checkpoint so that it can be used
during restoration to correctly initialize all doorbells. Additionally,
this commit also sets the hsa queue correctly during checkpoint-restoration

Change-Id: Ief3ef6dc973f70f27255234872a12c396df05d89

v-ramadas committed:
mem-ruby: Add Flush as a supported memory type in VIPERCoalescer

This commit adds flush as a recognized memory type in VIPERCoalescer.

Change-Id: I0f1b6f4518548e8e893ef681955b12a49293d8b4

v-ramadas committed:
mem-ruby: Add cache cooldown and warmup support to GPUCoalescer

The GPU Coalescer does not contain cache cooldown and warmup support.
This commit updates the coalsecer to support cache cooldown during flush
and warmup during checkpoint restore.

Change-Id: I5459471dec20ff304fd5954af1079a7486ee860a

v-ramadas committed:
mem-ruby: Add flush support to GPU_VIPER protocol

This commit adds flush support to the GPU VIPER coherence protocol. The
L1 cache will now initiate a flush request if the packet it receives
is of type RubyRequestType_FLUSH. During the flush process, the L1 cache
will a request to L2 if its in either V or I state. L2 will issue a
flush request to the directory if its cache line is in the valid
state before invalidating its copy. The directory, on receiving this
request, writes data to memory and sends an ack back to the L2. L2
forwards this ack back to the L1, which then ends the flush by calling
the write callback

Change-Id: I9dfc0c7b71a1e9f6d5e9e6ed4977c1e6a3b5ba46

v-ramadas committed:
mem-ruby: Update cache recorder to use GPUCoalescer port for GPUs

Previously, the cache recorder used the Sequencer to issue flush
requests and cache warmup requests. The GPU however uses GPUCoalescer to
access the cache, and not the Sequencer. This commit adds a GPUCoalescer
map to the cache recorder and uses it to send flushes and cache warmup
requests to any GPU caches in the system

Change-Id: I10490cf5e561c8559a98d4eb0550c62eefe769c9

abmerop commented on src/dev/amdgpu/pm4_packet_processor.cc:
```diff
@@ -1164,27 +1173,29 @@ PM4PacketProcessor::unserialize(CheckpointIn &cp)
         queues[id[i]]->wptr(wptr[i]);
         queues[id[i]]->ib(true);
         queues[id[i]]->wptr(ib_wptr[i]);
+        queues[id[i]]->rptr(ib_rptr[i]);
```
This whole thing seems kind of weird.  All three of these get called twice, which one is correct?  I will also check after this review... I think we need to be seeing if it's an IB (indirect buffer) or not first and using the appropriate read/write pointers and ib boolean

v-ramadas replied:
I updated this to write the appropriate read/write pointers to the queue based on the ib field value

abmerop commented on src/dev/amdgpu/pm4_packet_processor.cc:
```diff
@@ -1164,27 +1173,29 @@ PM4PacketProcessor::unserialize(CheckpointIn &cp)
     UNSERIALIZE_ARRAY(hqd_pq_control, num_queues);
 
-    for (int i = 0; i < num_queues; i++) {
+    for (int i = num_queues - 1; i >= 0; i--) {
```
Any reason why this is reversed?  It it necessary?

v-ramadas replied:
I ran into a tlb related issue when restoring, and wanted to check if it was because of the order in which I restore. Since this diagnosis was incorrect, this change is not needed. I have reverted it to the original version

abmerop commented on src/dev/amdgpu/pm4_packet_processor.cc:
```diff
@@ -1044,6 +1044,7 @@ PM4PacketProcessor::serialize(CheckpointOut &cp) const
     int num_queues = queues.size();
     Addr id[num_queues];
     Addr mqd_base[num_queues];
+    uint64_t mqd_read_index[num_queues];
```
This is fine for now -  I originally added these and you are following the style around it, I understand.  But at some point we should remove these VLAs (variable length arrays) because I don't think they are standard C++.  It happens to be OK with gcc/clang which gem5 supports but I think other compilers like MSVC will error out

abmerop commented on src/mem/ruby/system/GPUCoalescer.cc:
```diff
@@ -129,8 +137,16 @@ UncoalescedTable::updateResources()
             // Release the token
-            DPRINTF(GPUCoalescer, "Returning token seqNum %d\n", seq_num);
-            coalescer->getGMTokenPort().sendTokens(1);
+            if (!RubySystem::getWarmupEnabled()
+                    && !RubySystem::getCooldownEnabled()) {
```
For this and other places where we are checking if Ruby is in warmup, we should probably comment why the code is conditioned out

v-ramadas replied:
I added a few comments wherever Ruby checks if it is in warmup or cooldown.

abmerop reviewed (commented):
Thanks, this is great work so far!  It will be a huge feature to be able to checkpoint/restore between GPU kernels.  Everything looks correct to me, I just have a few recommendations below

v-ramadas committed:
dev-amdgpu: Remove duplicate writes to PM4 queue pointers

During checkpoint restoration, the unserialize() function writes rptr,
wptr, and indirect buffer rptr, wptr to PM4 queue's rptr, wptr fields.
This commit updates this to write only the relevant pointers to the
queue structure. If indirect buffers are used, then it writes only the
indirect buffer pointers to the queue. If they are not used, then it
writes rptr, wptr values to the queue.

Change-Id: Iedb25a726112e1af99cc1e7bc012de51c4ebfd45

abmerop commented:
Thanks for the updates, this looks good to me.  I guess we are just keeping it a draft while deciding how to handle the AVX save/restore issue?

v-ramadas commented:
I was thinking of either reverting the commit https://github.com/gem5/gem5/pull/113/commits/9acfc5a751c4df408b2fa9f73aa4512392c48c02 or updating the GPUFS README to ask the user to disable these fields before checkpointing until xsave/xrstor instructions are implemented (and the problem is fixed). What would you suggest?

abmerop commented:
I would just add a command line option to enable/disable AVX, then use that to set use_avx in system.py.  Please don't revert that commit as it will break DNNMark.

v-ramadas committed:
configs: Add option to disable AVX in GPUFS

GPUFS+KVM simulations automatically enable AVX. This commit adds a
command line option to disable AVX if its not needed for a GPUFS
simulation.

Change-Id: Ic22592767dbdca86f3718eca9c837a8e29b6b781

v-ramadas commented:
Thanks for the suggestion! I've pushed a new commit to this PR that adds an option to disable avx. I've also removed the WIP tag and marked the PR ready for review

abmerop reviewed (dismissed):
Looks good to me, thanks for all the changes!

mattsinc commented:
@v-ramadas : this is the error:

src/gpu-compute/gpu_dyn_inst.hh:42:10: fatal error: debug/GPUMem.hh: No such file or directory
   42 | #include "debug/GPUMem.hh"
   
Do things compile for you locally?

mattsinc commented on src/mem/ruby/protocol/GPU_VIPER-TCP.sm:
```diff
@@ -469,6 +474,24 @@ machine(MachineType:TCP, "GPU TCP (L1 Data Cache)")
+                              TCC_select_low_bit, TCC_select_num_bits));
+      out_msg.MessageSize := MessageSizeType:Data;
+      out_msg.Type := CoherenceRequestType:WriteFlush;
+      out_msg.InitialRequestTime := curCycle();
```
does this need to be curTick?

v-ramadas replied:
I set it to curCycle() since other actions that set out_msg.InitialRequestTime also set it to curCycle(). I think this field is used to calculate the latency after the request completes, which might not be important during flush.

mattsinc reviewed (commented):
We'll have to figure out why the error is happening, otherwise one small question

abmerop commented:
> We'll have to figure out why the error is happening, otherwise one small question

I think it is due to the ALL build not building the GPU, and the DebugFlags for GPUMem, etc. are guarded in src/gpu-compute/SConscript are guarded by the configuration variable BUILD_GPU.

I assume this means that all GPU patches will now fail tests if that's the case

v-ramadas commented:
> @v-ramadas : this is the error:
> 
> src/gpu-compute/gpu_dyn_inst.hh:42:10: fatal error: debug/GPUMem.hh: No such file or directory 42 | #include "debug/GPUMem.hh"
> 
> Do things compile for you locally?

Yes, I am able to compile and execute locally.

abmerop commented on src/mem/ruby/system/CacheRecorder.cc:
```diff
@@ -30,8 +30,11 @@
 
 #include "debug/RubyCacheTrace.hh"
+#include "mem/packet.hh"
+#include "mem/ruby/system/GPUCoalescer.hh"
```
To fix the ALL test, I think we need to wrap the GPUCoalescer.hh header with an `#ifdef BUILD_GPU` / `#endif`. 

For example: https://github.com/gem5/gem5/blob/stable/src/mem/ruby/profiler/Profiler.cc#L74

Hopefully this won't be needed if we go forward with #306

v-ramadas replied:
Thank you for pointing this out! I've added similar guards and I am able to compile both VEGA_X86 and X86 locally. Also verified that this does not break checkpointing.

v-ramadas committed:
mem-ruby: Add BUILD_GPU guard to ruby cooldown and warmup phases

Ruby was recently updated to support flushes and warmup for GPUs. Since
this support uses the GPUCoalescer, non-GPU builds face a compile time
issue. This is because GPU code is not built for non-GPU builds. This
commit addes "#if BUILD_GPU" guards around the GPU-related code in
common files like AbstractController.hh, CacheRecorder.*, RubySystem.cc,
GPUCoalescer.hh, and VIPERCoalescer.hh. This support allows GPU builds
to use flushing while non-GPU builds compile without problems

Change-Id: If8ee4ff881fe154553289e8c00881ee1b6e3f113

powerjg commented on src/mem/ruby/system/CacheRecorder.cc:
```diff
@@ -143,13 +180,27 @@ CacheRecorder::enqueueNextFetchRequest()
+#endif
             assert(m_sequencer_ptr != NULL);
+#if BUILD_GPU
+            if (m_coal_ptr == NULL)
+                m_sequencer_ptr->makeRequest(pkt);
```
```suggestion
            if (m_coal_ptr == NULL) {
                m_sequencer_ptr->makeRequest(pkt);
            }
```

mattsinc commented on src/mem/ruby/system/CacheRecorder.cc:
```diff
@@ -54,14 +57,29 @@ CacheRecorder::CacheRecorder()
 {
 }
 
+#if BUILD_GPU
```
shouldn't this be #ifdef instead of #if?

v-ramadas replied:
At the start of compile time, configs/build_gpu.hh is generated with a value of either 0 or 1 (if VEGA_X86) for  BUILD_GPU. #ifdef always returns true since BUILD_GPU is defined in either case. I used #if to differentiate between its possible values 0 and 1.

mattsinc replied:
I am not sure I understand the distinction, but ok.

powerjg replied:
I'm unconvinced this was a safe thing to do. It is bad practice to rely on magic values like this! 

In fact, I'm pretty sure that #69 will break this change.

I'm quite concerned about adding more build-time guard flags. We are trying to move away from configuring gem5 at build time (see change for multi-ISA build (already merged) and multi-Ruby build (not yet merged)).

mattsinc replied:
Do you have a recommendation about an alternative then?

powerjg replied:
There's potential problems with this code that I would like to see explained:
1. It is bad practice to use magic numbers like the fact that BUILD_GPU is 0 when not VEGA_X86 (or something like that). If we use a compile-time define, we must use something that is explicit and not implicit.
2. We want to avoid compile-time defines. I would need to spend time understanding the problem and trying to come up with a different solution. It would be great if @v-ramadas would try to come up with a different solution to the problem.

What *is* the problem this is `#if` is trying to solve?

mattsinc replied:
I am not sure I understand the implicit/explicit part, because BUILD_GPU is explicitly set to 1 in the case you described.  I assume you mean it is not explicitly set to 0 in other builds?

Ultimately from @abmerop above, this is the problem:

"I think it is due to the ALL build not building the GPU, and the DebugFlags for GPUMem, etc. are guarded in src/gpu-compute/SConscript are guarded by the configuration variable BUILD_GPU.

I assume this means that all GPU patches will now fail tests if that's the case."

The only options I can think of are:

1. Don't guard the GPU DebugFlags with BUILD_GPU -- I guess there is no harm in them being accessible in a build that is not for the GPU?
2. Have ALL build the GPU -- not sure if there is a Board/stdlib related reason for this though?
3. Use something like ifdef's and explicitly set BUILD_GPU to false in things that aren't GCN3_X86/VEGA_X86.

mattsinc commented on src/mem/ruby/system/CacheRecorder.cc:
```diff
@@ -81,6 +99,9 @@ CacheRecorder::~CacheRecorder()
         m_uncompressed_trace = NULL;
     }
     m_seq_map.clear();
+#if BUILD_GPU
```
same here

mattsinc reviewed (commented):
If if/ifdef behave the same, maybe fine.  But one question remains.

powerjg commented:
I had half reviewed this, but got distracted by other things. I'm quite concerned about the `#if` change.

mattsinc commented:
> I had half reviewed this, but got distracted by other things. I'm quite concerned about the `#if` change.

We can revert the commit to give you more time to review -- I am just not sure how to revert.

powerjg commented:
![image](https://github.com/gem5/gem5/assets/109596/1cf700de-4525-485f-8c01-151e2f085b2b)

You should be able to click the revert button. However, I would rather not add more commit spam... If we can (quickly) resolve this with a follow up change, that would be best.

mattsinc commented:
> > We'll have to figure out why the error is happening, otherwise one small question
> 
> I think it is due to the ALL build not building the GPU, and the DebugFlags for GPUMem, etc. are guarded in src/gpu-compute/SConscript are guarded by the configuration variable BUILD_GPU.
> 
> I assume this means that all GPU patches will now fail tests if that's the case

@powerjg : this is the error the #if was fixing -- the compilation failed because the added debug flag is guarded by BUILD_GPU.

abmerop commented:
Implementing #403 should get rid of most (all?) of the new ones in the patch.  We could remove the other ones in various places of the code if we go forward with #306.

mattsinc commented:
> Implementing #403 should get rid of most (all?) of the new ones in the patch. We could remove the other ones in various places of the code if we go forward with #306.

You're right @abmerop , #403 is likely the solution here.  My above description is also not precise enough -- the real issue is this separate GPUCoalescer map we need without #403.

@powerjg I guess the choice then revert this or let it be in/merged until #403 is done.  @v-ramadas is prioritizing #403 as the next fix, but it will still be a few days.

v-ramadas commented:
I've created a PR (#448) to address #403 and remove the #if guards
